File: src/68_01_CommonParentInSearchTree/68_01_CommonParentInSearchTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

import os
import sys

sys.path.append(os.path.split(sys.path[0])[0])


from Utilities import binary_tree
from Utilities.binary_tree import BinaryTreeNode as TreeNode


class Solution:
    def lowestCommonAncestor(
        self, root: "TreeNode", p: "TreeNode", q: "TreeNode", use_iter: bool = False
    ) -> "TreeNode":
        if not root:
            return None

        if use_iter:
            return self.withIter(root, p, q)
        else:
            return self.withRecur(root, p, q)

    def withIter(self, root: "TreeNode", p: "TreeNode", q: "TreeNode"):
        """迭代"""
        # 保证p.val<q.val，减少后续判断
        if p.val > q.val:
            p, q = q, p
        while root:
            if root.val < p.val and root.val < q.val:
                root = root.right
            elif root.val > p.val and root.val > q.val:
                root = root.left
            else:
                break
        return root

    def withRecur(self, root: "TreeNode", p: "TreeNode", q: "TreeNode"):
        """递归"""
        if root.val < p.val and root.val < q.val:
            return self.withRecur(root.right, p, q)
        if root.val > p.val and root.val > q.val:
            return self.withRecur(root.left, p, q)
        return root


class Test:
    def __init__(self) -> None:
        self.solution = Solution()

    def checkResult(
        self,
        testName: str,
        root: TreeNode,
        p: TreeNode,
        q: TreeNode,
        expected: TreeNode,
    ) -> None:
        text = testName
        result = self.solution.lowestCommonAncestor(root, p, q)
        if (not expected and not result) or (
            expected and result and expected.val == result.val
        ):
            text += " passed."
        else:
            text += " failed."
        print(text)

    #      6
    #    /   \
    #   2     8
    #  / \    / \
    # 0   4  7   9
    #    / \
    #   3   5
    # p=0, q=5, expected=2
    def Test1(self):
        vals = [6, 2, 8, 0, 4, 7, 9, 3, 5]
        nodes = binary_tree.createNodesVec(vals)

        binary_tree.connectNodesIndex(nodes, 0, 1, 2)
        binary_tree.connectNodesIndex(nodes, 1, 3, 4)
        binary_tree.connectNodesIndex(nodes, 2, 5, 6)
        binary_tree.connectNodesIndex(nodes, 4, 7, 8)

        self.checkResult("Test1", nodes[0], nodes[3], nodes[8], nodes[1])

    #               5
    #              /
    #             4
    #            /
    #           3
    #          /
    #         2
    #        /
    #       1
    # p=1, q=3, expected=4
    def Test2(self):
        vals = [5, 4, 3, 2, 1]
        nodes = binary_tree.createNodesVec(vals)

        binary_tree.connectNodesIndex(nodes, 0, 1, -1)
        binary_tree.connectNodesIndex(nodes, 1, 2, -1)
        binary_tree.connectNodesIndex(nodes, 2, 3, -1)
        binary_tree.connectNodesIndex(nodes, 3, 4, -1)

        self.checkResult("Test2", nodes[0], nodes[4], nodes[2], nodes[2])

    # 输入None
    def Test3(self):
        self.checkResult("Test3", None, None, None, None)

    def runTest(self):
        self.Test1()
        self.Test2()
        self.Test3()


if __name__ == "__main__":
    test = Test()
    test.runTest()

File: src/68_01_CommonParentInSearchTree/68_01_CommonParentInSearchTree.cpp
// 面试题68： 二叉搜索树的最近公共祖先
// 题目：给定一个二叉搜索树, 找到该树中两个指定节点的最近公共祖先。
// 所有节点的值都是唯一的, p、q 为不同节点且均存在于给定的二叉搜索树中

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 * };
 */

#include <iostream>
#include <string>
#include <vector>

#include "../Utilities/BinaryTree.h"

using namespace std;

class Solution {
   public:
    // 如果p和q在root两侧，那么root就是它们的共同祖先
    TreeNode* lowestCommonAncestor( TreeNode* root, TreeNode* p, TreeNode* q,
                                    bool use_iter = false ) {
        if ( !root ) return nullptr;
        if ( use_iter )
            return withIter( root, p, q );
        else
            return withRecur( root, p, q );
    }

   private:
    // 迭代
    TreeNode* withIter( TreeNode* root, TreeNode* p, TreeNode* q ) {
        // 保证p.val<q.val，减少后续判断
        if ( p->val > q->val ) std::swap( p, q );
        while ( root ) {
            if ( root->val < p->val && root->val < q->val )  // p,q都在右子树
                root = root->right;
            else if ( root->val > p->val && root->val > q->val )  // 都在左子树
                root = root->left;
            else  // 如果分别在两侧，说明找到
                break;
        }
        return root;
    }

    // 递归
    TreeNode* withRecur( TreeNode* root, TreeNode* p, TreeNode* q ) {
        if ( root->val < p->val && root->val < q->val )
            return withRecur( root->right, p, q );
        if ( root->val > p->val && root->val > q->val )
            return withRecur( root->left, p, q );
        return root;
    }
};

class Test {
   public:
    void checkResult( const string testName, TreeNode* root, TreeNode* p,
                      TreeNode* q, TreeNode* expected ) {
        cout << testName;
        TreeNode* result = solution.lowestCommonAncestor( root, p, q );
        if ( ( !expected && !result ) ||
             ( expected && result && expected->val == result->val ) )
            cout << " passed.\n";
        else
            cout << " failed.\n";
    }

    //      6
    //    /   \
    //   2     8
    //  / \    / \
    // 0   4  7   9
    //    / \
    //   3   5
    // p=0, q=5, expected=2
    void Test1() {
        ValVec vals{ 6, 2, 8, 0, 4, 7, 9, 3, 5 };
        TreeNodeVec nodes = createTreeNodesVec( vals );

        TreeIndexVec idxs{
            { 0, 1, 2 },
            { 1, 3, 4 },
            { 2, 5, 6 },
            { 4, 7, 8 },
        };
        connectTreeNodesIndexVec( nodes, idxs );

        checkResult( "Test1", nodes[ 0 ], nodes[ 3 ], nodes[ 8 ], nodes[ 1 ] );
        destroyTree( nodes[ 0 ] );
    }

    //               5
    //              /
    //             4
    //            /
    //           3
    //          /
    //         2
    //        /
    //       1
    // p=1, q=3, expected=4
    void Test2() {
        ValVec vals{ 5, 4, 3, 2, 1 };
        TreeNodeVec nodes = createTreeNodesVec( vals );

        TreeIndexVec idxs{
            { 0, 1, -1 },
            { 1, 2, -1 },
            { 2, 3, -1 },
            { 3, 4, -1 },
        };
        connectTreeNodesIndexVec( nodes, idxs );

        checkResult( "Test2", nodes[ 0 ], nodes[ 4 ], nodes[ 2 ], nodes[ 2 ] );
        destroyTree( nodes[ 0 ] );
    }

    // 输入nullptr
    void Test3() { checkResult( "Test3", nullptr, nullptr, nullptr, nullptr ); }

    void runTest() {
        Test1();
        Test2();
        Test3();
    }

   private:
    Solution solution;
};

int main() {
    Test test;
    test.runTest();
    return 0;
}